### PR TITLE
Category

### PR DESCRIPTION
**Pull request by DenisBogusevich**

### CODE REVIEW

File: src/main/java/onlinebookstore/controler/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        categoryService.deleteById(id);
+    }
+
+    @GetMapping("/{id}/books")
+    @PreAuthorize("hasAnyRole('USER','ADMIN')")
+    @ResponseStatus(HttpStatus.OK)
+    public List<BookDtoWithoutCategoryIds> getBooksByCategoryId(@PathVariable Long id) {
+        return bookService.getBooksByCategoryId(id);
+    }
+
```
Comment (Elena-Bruyako) [inline src/main/java/onlinebookstore/controler/CategoryController.java]:
```suggestion
```

File: src/main/java/onlinebookstore/controler/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    @PreAuthorize("hasRole('ADMIN')")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void deleteCategoryById(@PathVariable Long id) {
+        categoryService.deleteById(id);
+    }
+
+    @GetMapping("/{id}/books")
+    @PreAuthorize("hasAnyRole('USER','ADMIN')")
+    @ResponseStatus(HttpStatus.OK)
+    public List<BookDtoWithoutCategoryIds> getBooksByCategoryId(@PathVariable Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/onlinebookstore/controler/CategoryController.java]:
```suggestion
    public List<BookDtoWithoutCategoryIds> getBooksByCategoryId(@PathVariable @Positive Long id) {
```

File: src/main/java/onlinebookstore/controler/CategoryController.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    @ResponseStatus(HttpStatus.OK)
+    @PreAuthorize("hasAnyRole('USER','ADMIN')")
+    public CategoryDto getCategoryById(@PathVariable Long id) {
+        return categoryService.findById(id);
+    }
+
+    @PutMapping("/{id}")
+    @PreAuthorize("hasRole('ADMIN')")
+    @ResponseStatus(HttpStatus.OK)
+    public CategoryDto updateCategory(@PathVariable Long id,
```
Comment (Elena-Bruyako) [inline src/main/java/onlinebookstore/controler/CategoryController.java]:
```suggestion
    public CategoryDto updateCategory(@PathVariable @Positive Long id,
```

File: src/main/java/onlinebookstore/controler/CategoryController.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+                                      @RequestBody CategoryRequestDto categoryDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/onlinebookstore/controler/CategoryController.java]:
```suggestion
                                      @RequestBody @Valid CategoryRequestDto categoryDto) {
```

File: src/main/java/onlinebookstore/controler/CategoryController.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        return categoryService.update(id, categoryDto);
+    }
+
+    @DeleteMapping("/{id}")
+    @PreAuthorize("hasRole('ADMIN')")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void deleteCategoryById(@PathVariable Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/onlinebookstore/controler/CategoryController.java]:
```suggestion
    public void deleteCategoryById(@PathVariable @Positive Long id) {
```

File: src/main/java/onlinebookstore/controler/CategoryController.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
```
Comment (Elena-Bruyako) [inline src/main/java/onlinebookstore/controler/CategoryController.java:31]: Please add Swagger documentation here

File: src/main/java/onlinebookstore/controler/CategoryController.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+@RequiredArgsConstructor
+@RequestMapping("/categories")
```
Comment (Elena-Bruyako) [inline src/main/java/onlinebookstore/controler/CategoryController.java]: `@Validated`

File: src/main/java/onlinebookstore/controler/CategoryController.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+public class CategoryController {
+    private final CategoryService categoryService;
+    private final BookService bookService;
+
+    @PostMapping
+    @PreAuthorize("hasRole('ADMIN')")
+    @ResponseStatus(HttpStatus.CREATED)
+    public CategoryDto createCategory(@RequestBody CategoryRequestDto categoryRequestDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/onlinebookstore/controler/CategoryController.java]:
```suggestion
    public CategoryDto createCategory(@RequestBody @Valid CategoryRequestDto categoryRequestDto) {
```

File: src/main/java/onlinebookstore/controler/CategoryController.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        return categoryService.save(categoryRequestDto);
+    }
+
+    @GetMapping
+    @PreAuthorize("hasAnyRole('USER','ADMIN')")
+    @ResponseStatus(HttpStatus.OK)
+    public List<CategoryDto> getAll() {
```
Comment (Elena-Bruyako) [inline src/main/java/onlinebookstore/controler/CategoryController.java]:
```suggestion
    public List<CategoryDto> getAll(@ParameterObject @PageableDefault Pageable pageable) {
```

File: src/main/java/onlinebookstore/controler/CategoryController.java [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        return categoryService.findAll();
+    }
+
+    @GetMapping("/{id}")
+    @ResponseStatus(HttpStatus.OK)
+    @PreAuthorize("hasAnyRole('USER','ADMIN')")
+    public CategoryDto getCategoryById(@PathVariable Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/onlinebookstore/controler/CategoryController.java]:
```suggestion
    public CategoryDto getCategoryById(@PathVariable @Positive Long id) {
```

File: src/main/java/onlinebookstore/dto/book/CreateBookRequestDto.java [hunk 11]
```diff
@@ -15,5 +17,7 @@ public record CreateBookRequestDto(
         @PositiveOrZero
         BigDecimal price,
         String description,
-        String coverImage) {
+        String coverImage,
+        @NotEmpty
+        List<Long> categoryIds) {
```
Comment (Elena-Bruyako) [inline src/main/java/onlinebookstore/dto/book/CreateBookRequestDto.java]:
```suggestion
        List<@Positive Long> categoryIds) {
```

File: src/main/java/onlinebookstore/mapper/CategoryMapper.java [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+import org.mapstruct.NullValuePropertyMappingStrategy;
+
+@Mapper(componentModel = "spring",
+        injectionStrategy = InjectionStrategy.CONSTRUCTOR,
+        nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS,
+        implementationPackage = "<PACKAGE_NAME>.impl")
+public interface CategoryMapper {
+    CategoryDto toDto(Category category);
+
+    Category toEntity(CategoryRequestDto categoryDto);
```
Comment (Elena-Bruyako) [inline src/main/java/onlinebookstore/mapper/CategoryMapper.java]:
```suggestion
    Category toCategory(CategoryRequestDto categoryDto);
```

File: src/main/java/onlinebookstore/repository/BookRepository.java [hunk 13]
```diff
@@ -1,9 +1,12 @@
 package onlinebookstore.repository;
 
+import java.util.List;
 import onlinebookstore.entity.Book;
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.stereotype.Repository;
 
 @Repository
 public interface BookRepository extends JpaRepository<Book, Long> {
+
```
Comment (Elena-Bruyako) [inline src/main/java/onlinebookstore/repository/BookRepository.java]:
```suggestion
```

File: src/main/java/onlinebookstore/service/impl/CategoryServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    @Override
+    public CategoryDto save(CategoryRequestDto categoryDto) {
+        return categoryMapper.toDto(categoryRepository.save(categoryMapper.toEntity(categoryDto)));
+    }
+
+    @Override
+    public CategoryDto findById(Long id) {
+        return categoryRepository.findById(id)
+                .map(categoryMapper::toDto)
+                .orElseThrow();
```
Comment (Elena-Bruyako) [inline src/main/java/onlinebookstore/service/impl/CategoryServiceImpl.java]:
```suggestion
                .orElseThrow(() -> new EntityNotFoundException(.......);
```

File: src/main/resources/db/changelog/changes/06-create-category-table.yaml [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+                    nullable: false
+              - column:
+                  name: name
+                  type: VARCHAR(255)
+                  constraints:
+                    unique: true
+                    nullable: false
+              - column:
+                  name: description
+                  type: VARCHAR(2000)
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/06-create-category-table.yaml]:
```suggestion
                  type: VARCHAR(255)
```
2000 - it's too big size.

File: src/main/resources/db/changelog/changes/06-create-category-table.yaml [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+              - column:
+                  name: is_deleted
+                  type: boolean
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/06-create-category-table.yaml:30]: add empty line at the end of file
Check all changelogs

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Lets improve your solution!

**Review by liliia-ponomarenko — APPROVED** [review]
Good job! But fix the conflict before merging ;)

**Pull request closed without merging**